### DISCUSSION TITLE
Building permissions in superset

### DISCUSSION THREAD

**Discussion by maya-harel** [apache/superset#19332] [Q&A / Help]
Hi! I am integrating superset into my application, and I want to understand how to build permissions for users

I want to give users permission per schema (data is on AWS Athena) and I want to allow only creation of Charts and Dashboards (in the future also SQL queries for some of the users)

And (most important) I want NOT to allow exporting and downloading of any kind (not charts/dashboards/datasets/databases/queries)
(And also as a bonus I don't want to allow to copy links to charts/dashboards etc)

What I did up until now was to create a role for each schema, and give it permissions for `schema_access` for the specified schema, but I see that users are able to export data with this role

I want to understand which role I need to build for my users 

Thanks in advance!

**Comment by srinisubramanian** [comment]
I guess you should look into the permissions for the role via Roles -> Edit.  There are some permissions like Export Dashboard, Save etc that you may have to remove.  Best is to copy the Alpha or Gamma roles and customize them

**Reply by maya-harel** [reply]
Hi :) 
Thank you for the suggestion, I created a custom role, copied permissions to it from the Gamma role, and customized it.

There are a few specific permissions I was not able to block/customize through the roles, if you have any insights:

- I was able to remove permission to export charts, but this only disables the option to export as CSV - the option to export as JSON still stands (cannot find a way to remove this using the roles)
- Option to share a link to a chart - is there any way to remove this functionality? also to share by email

I am also trying to understand if there is a pattern for all exporting through superset (<superset url>/api/v1/dashboard/export), because when I export from a chart from example the URL is <superset url>/api/v1/dashboard/data
(I am wondering if a can block this on an infrustructure level)

**Reply by srinisubramanian** [reply]
I guess this might be a little trial and error.  An alternative is to start with an Empty role and enabling just the Dashboard/Chart/Specific data access and try.  That might be an easier option to provide the basic/minimum access?

**Reply by srinisubramanian** [reply]
And for export there are a number of end points depending on the object being exported.  Check here: https://superset.apache.org/docs/api/ and under most objects like dashboard, dataset, chart etc you will find an export

**Reply by maya-harel** [reply]
I have tried blocking `<superset url>/api/v1/*/export`, but when exporting through the UI, the URL is `<superset url>/api/v1/*/data`, but I cannot block this because than I am not able to see the chart at all

**Comment by nagarajmmu** [comment]
We are also looking for customizing the permissions, if any one achieved permission customization, please let us know. thanks in advance
